server/src/main/java/org/apache/druid/discovery/DruidNodeDiscoveryProvider.java

```diff
@@ -235,6 +235,28 @@ public void nodeViewInitialized()
           }
         }
       }
+
+      @Override
+      public void nodeViewInitializedTimedOut()
+      {
+        synchronized (lock) {
+          if (uninitializedNodeRoles == 0) {
+            log.error("Unexpected call of nodeViewInitializedTimedOut()");
+            return;
+          }
+          uninitializedNodeRoles--;
+          if (uninitializedNodeRoles == 0) {
+            for (Listener listener : listeners) {
+              try {
+                listener.nodeViewInitializedTimedOut();
+              }
+              catch (Exception ex) {
+                log.error(ex, "Listener[%s].nodeViewInitializedTimedOut() threw exception. Ignored.", listener);
```

Comment:
This logic looks very similar to the `nodeViewInitialized()` logic. Better to extract the shared stuff to a private method like `nodeViewInitialized(boolean timedOut)`.